lib/tests/maintainers.nix

```
@@ -20,16 +20,20 @@ let
         ];
       }).config;
 
-      checkGithubId = lib.optional (checkedAttrs.github != null && checkedAttrs.githubId == null) ''
+      checks = lib.optional (checkedAttrs.github != null && checkedAttrs.githubId == null) ''
         echo ${lib.escapeShellArg (lib.showOption prefix)}': If `github` is specified, `githubId` must be too.'
         # Calling this too often would hit non-authenticated API limits, but this
         # shouldn't happen since such errors will get fixed rather quickly
         info=$(curl -sS https://api.github.com/users/${checkedAttrs.github})
         id=$(jq -r '.id' <<< "$info")
         echo "The GitHub ID for GitHub user ${checkedAttrs.github} is $id:"
         echo -e "    githubId = $id;\n"
+      '' ++ lib.optional (checkedAttrs.email == null && checkedAttrs.github == null && checkedAttrs.matrix == null) ''
```

```suggestion
      '' ++ lib.optional (checkedAttrs.email == null && checkedAttrs.matrix == null) ''
```

There's no way to reach out to arbitrary GitHub users

---

Can't you just @ them?

---

Note that this change would make 62 maintainers invalid.

---

You can't reach out to GitHub users privately though (unless I guess you create a private repository, invite the desired users to it, then create an issue and ping them). Also when you see a GitHub users profile and they have no public email address this might be their way of saying that they don't want to be contacted privately.

I guess the question is: Do we want nixpkgs maintainers to be contactable privately? While that would be nice, I don't think it's actually needed. How about just requiring them to have a GitHub account instead, ideally with the requirement that that very GitHub user makes the PR that adds them as a maintainer. Since nixpkgs is entirely public, there shouldn't be a reason to contact people privately. If you have a question regarding a package you can file an issue and ping the maintainers.

---

> Do we want nixpkgs maintainers to be contactable privately?

Hell no. We already have a [security team](https://nixos.org/community/teams/security.html) for disclosures.

> How about just requiring them to have a GitHub account instead

I'm kinda :| on that. Ideally you should be able to submit patches to nixpkgs by email. I guess there's little sense in having only a matrix ID specified though.

---

I see, I think just being reachable by any public means is good enough.

Considering that we can't prevent people from being entirely unreachable anyways, we might as well assume that all users do want to be reachable. Users that don't want to be reachable can always find a way (just enter invalid/wrong fields or ignore all messages). And in that case this check here does make sense as is and we can't do any better.

---

Well, verifying the fields in some way would at least prevent mistakes and impersonations (using a different-but-valid username/email), so I think that would still have some value, but that's for another day.

---

What happens if someones github account gets deleted and they have no other contact mean?
If we don't know anything else about them, then we would need to delete their maintainer entry.

---

Good point. Maybe we could then set their GitHub ID to `null` or something to signify that it's invalid?

---

I'd say we should remove the maintainer entry in that case

---

I know that it's very unpractical, but someone could theoretically contribute without an account by submitting patches or some other form of (possibly human) proxy. I really dislike the fact the GitHub gets to decide who can or can't contribute to Nixpkgs, say because their country got sanctioned by the USA or they broke some kind of term of service.

---

I would suggest that in such case they need to find a person willing to do the proxy part and also handles review comments, questions, bugs, etc.

---

Turns out there's only one person without an associated GitHub account:
```nix
nix-repl> :p builtins.filter (attr: ! attr ? github) (builtins.attrValues (import maintainers/maintainer-list.nix))                      
[
  { email = "davide.peressoni@tuta.io"; githubId = 3186857; matrix = "@dpd-:matrix.org"; name = "Davide Peressoni"; }
  { email = "loveisgrief@tuta.io"; keys = [ { fingerprint = "9847 4F48 18C6 4E0A F0C5  3529 E96D 1EDF A053 45EB"; } ]; name = "LoveIsGrief"; }
]
```

And that looks like it might be a mistake when @LoveIsGrief added themselves in https://github.com/NixOS/nixpkgs/pull/209769 (there is an obviously associated GitHub account)

While this makes this check here more strict, it wouldn't be a problem because of that

---

I'm not sure I understand the mistake. The file header says that email and name are required https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix#L4 . I followed what it says.

IMO, this PR isn't complete if `maintainers-list.nix` keep the header with the email+name requirements, if the email is rendered optional.

---

Thanks for pointing that out, that comment should indeed be updated to stay in sync with whatever check we decide to settle on.

Regardless, we should also mention that setting a GitHub ID allows ofborg to request your review automatically when a package you maintain is modified.

---

Regardless of the other ideas here, I think this check in this PR is good as is. The only thing to do is to update the comment in the maintainer listing

---

> Turns out there's only one person without an associated GitHub account:

That's because they have all been removed in https://github.com/NixOS/nixpkgs/pull/178656

Famously, there was @volth that received a full damnatio memoriae by github but contributed patches using the discourse forum, I think.